じゃんけん課題

yusukedayo opened the pull request:
確認よろしくお願いします。

yusukedayo committed:
じゃんけん課題

確認よろしくお願いします。

miketa-webprgr commented:
要件

```
$ ruby main.rb

何本勝負？(press 1 or 3 or 5)
3本勝負を選びました

1本目
じゃんけん…(press g or c or p)
CPU…パー
あなた…チョキ
勝ち！
1勝0敗

2本目
じゃんけん…(press g or c or p)
CPU…パー
あなた…グー
負け！
1勝1敗

3本目
じゃんけん…(press g or c or p)
CPU…パー
あなた…パー
あいこで...(press g or c or p)
CPU…パー
あなた…チョキ

勝ち！
2勝1敗


結果
2勝1敗であなたの勝ち
```

miketa-webprgr commented:
このプログラムの実行結果

```
何本勝負？(数字を入力してください 1 or 3 or 5)
5
5本勝負
5本勝負を選びました。
じゃんけん（g, c, pから選択してください）
g
CPU g
あなた g
あいこ
じゃんけん（g, c, pから選択してください）
g
CPU p
あなた g
あなたの負け
0勝1敗

じゃんけん（g, c, pから選択してください）
p
CPU c
あなた p
あなたの負け
0勝2敗

じゃんけん（g, c, pから選択してください）
c
CPU p
あなた c
あなたの勝ち
1勝2敗

じゃんけん（g, c, pから選択してください）
c
CPU p
あなた c
あなたの勝ち
2勝2敗

じゃんけん（g, c, pから選択してください）
g
CPU p
あなた g
あなたの負け
2勝3敗

結果 2勝3であなたの負け

miketa-webprgr commented:
100%合わせろというのは細かすぎるかもしれないし、課題の趣旨とは違うかもしれないですが、
「CPU…パー」といった形で表示するといった要件あたりは満たしたほうがよいかと思います！

miketa-webprgr commented on jyanken.rb:
```diff
@@ -0,0 +1,77 @@
+
+# じゃんけん本体
+def jyanken
+    def battle
```
メソッドの中で、新しいメソッドを定義しないでください！（battleメソッドはjyankenメソッドの外に出す）

miketa-webprgr commented on jyanken.rb:
```diff
@@ -0,0 +1,77 @@
+    next_game = true
+    while next_game do
+        next_game = battle
+    end
```
こんな感じに書き換えられそうです。

```ruby
while battle do
  battle
end
```

miketa-webprgr replied:
> while文をご指摘頂いたように変更するとあいこ処理の数だけ勝負数も増えてしまうという不具合が発生してしまう

あ、あれ・・・
すみません・・・

miketa-webprgr commented:
ざっと見ました！

まずは、要件通り動かすのが大事になってくるかと思うので、もう少し頑張ってみてください！

あとは、実装し終わった後、他の方の課題を見て勉強するのが、特にじゃんけんの課題についてはすごくためになるので、そこはぜひやってみてください！（共通化するだとか、冗長に書かないだとか、そうしたところをまず学んでいくとよいかと思います！）

miketa-webprgr commented on jyanken.rb:
```diff
@@ -0,0 +1,77 @@
+# 勝敗回数を記録
+@result = [0, 0]
```
配列で勝敗を管理するというのは面白いアイデアですね！
私はwin, loseという変数にそれぞれ格納した記憶がありますが、配列というのはいい発想だと思います！

yusukedayo replied:
ほめてもらえると励みになります！ありがとうございます。
他の方のやり方も研究してたくさん学んでいきます！

yusukedayo committed:
修正版

while文をご指摘頂いたように変更するとあいこ処理の数だけ勝負数も増えてしまうという不具合が発生してしまうため
それ以外の部分を修正したものをとりあえず提出させて貰いました。
while文に関して自分で原因を考えてみたのですが違いが見つけられなかったのでアドバイス頂きたいです。
ちなみにwhile文に関しては以下のサイトを参考にして作成しました。
https://qiita.com/aiorange19/items/dff465f9e7af5e996f13

yusukedayo commented:
> 100%合わせろというのは細かすぎるかもしれないし、課題の趣旨とは違うかもしれないですが、 「CPU…パー」といった形で表示するといった要件あたりは満たしたほうがよいかと思います！

ご指摘ありがとうございます。
ハッシュを活用して要件通りの出力になるよう訂正しました！
```ruby
hands = {g: "グー", c:"チョキ", p:"パー"}
    user = gets.chomp
    puts "CPU… #{hands[cpu.to_sym]}"
```

yusukedayo committed:
修正#2

・whiel文をloop文に変更しました。
・何度やっても勝てない原因は前回修正した手の出し方を変更したのに勝敗条件の方では変更されていなかったからでした。
なので、手の出し方に合わせて勝敗条件の方も修正しました。
